hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java

```diff
@@ -144,13 +177,33 @@ public void stop() {
     if (executorService != null) {
       executorService.shutdown();
     }
-    if (hiveLocalDir != null) {
-      hiveLocalDir.delete();
-    }
     if (baseHandler != null) {
       baseHandler.shutdown();
     }
     METASTORE_THREADS_SHUTDOWN.invoke();
+    HMS_HANDLER_THREAD_LOCAL_CONF.remove();
+    HMS_HANDLER_THREAD_LOCAL_TXN.remove();
```

Comment:
What are you doing here?

Reply:
I'm trying to remove these thread local fields in HMSHandler, so that they won't get reused in follow-up tests.

Reply:
`HMSHandler.threadLocalConf` is supposed to be removed here which is called in the `shutdown`:
https://github.com/apache/hive/blob/branch-2.3/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java#L7287

You might be right that the `threadLocalTxn` is not cleaned up, but I would guess this does not affect our test yet, as we do not test transactional tables.

Reply:
I think spark 2.4 depends on `org.spark-project.hive:hive-exec:1.2.1.spark2` which is based on hive-1.2.1 and which doesn't clear `threadLocalConf ` in HMSHandler::shutdown method, right?

Reply:
Regarding `threadLocalTxn`, it's used when HMSHandler tries to acquire locks, so I think it also needs to be cleaned.

Reply:
I do not think that the threads get reused between tests. If ThreadLocals would be problematic, we should have faced the issues much sooner. Do we have tests where a TestHiveMetastore instance is created/recreated between test methods?

Reply:
I guess whether threads are reused depends on the testing framework. I can revert the clean up code to reproduce the test failures (did that earlier but seems CI report is lost after force push :slightly_frowning_face: )

Reply:
IMHO, reproducing the specific issue would be the best. And only fix the specific issue. Fixing everything would clutter the code seriously, and nobody would understand why all of these fixes are there. (I am hoping here that we would not face the other issues in the near future)

Reply:
I would be surprised if the tests would use 1.2.1 HMSHandler. When the error happens the stack traces shows this:
```
    	at org.apache.hadoop.hive.metastore.txn.TxnHandler.cleanupRecords(TxnHandler.java:1888)
    	at org.apache.hadoop.hive.metastore.AcidEventListener.onDropDatabase(AcidEventListener.java:51)
    	at org.apache.hadoop.hive.metastore.MetaStoreListenerNotifier$13.notify(MetaStoreListenerNotifier.java:69)
```
This code is not present in 1.2.1 Hive

Reply:
I meant spark 2.4 relies on hive 1.2.1. But in the latest CI, spark2 passed. I think it's not deterministic due to how tests are scheduled and how JVMs are reused. Seems we can force the problem to reproduce. E.g. I added this test case in `TestIdentityPartitionData` in spark 2.4 to do a restart of hive metastore:
```java
  @Test
  public void test() throws Exception {
    testFullProjection();
    stopMetastoreAndSpark();
    startMetastoreAndSpark();
    testFullProjection();
  }
```
And run the test with `./gradlew -DsparkVersions=2.4 :iceberg-spark:iceberg-spark2:test --tests org.apache.iceberg.spark.source.TestIdentityPartitionData`. The test would fail with:

```
org.apache.iceberg.spark.source.TestIdentityPartitionData > test[format = parquet, vectorized = false] FAILED
    java.lang.RuntimeException: Cannot start TestHiveMetastore
        at org.apache.iceberg.hive.TestHiveMetastore.start(TestHiveMetastore.java:134)
        at org.apache.iceberg.hive.TestHiveMetastore.start(TestHiveMetastore.java:97)
        at org.apache.iceberg.spark.SparkTestBase.startMetastoreAndSpark(SparkTestBase.java:57)
        at org.apache.iceberg.spark.source.TestIdentityPartitionData.test(TestIdentityPartitionData.java:143)

        Caused by:
        javax.jdo.JDOException: Exception thrown when executing query
        NestedThrowables:
        java.sql.SQLException: Container 2,177 not found.
            at org.datanucleus.api.jdo.NucleusJDOHelper.getJDOExceptionForNucleusException(NucleusJDOHelper.java:596)
            at org.datanucleus.api.jdo.JDOQuery.execute(JDOQuery.java:252)
            at org.apache.hadoop.hive.metastore.ObjectStore.getMRole(ObjectStore.java:3531)
            at org.apache.hadoop.hive.metastore.ObjectStore.addRole(ObjectStore.java:3221)
            at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
            at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
            at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
            at java.lang.reflect.Method.invoke(Method.java:498)
            at org.apache.hadoop.hive.metastore.RawStoreProxy.invoke(RawStoreProxy.java:114)
            at com.sun.proxy.$Proxy16.addRole(Unknown Source)
            at org.apache.hadoop.hive.metastore.HiveMetaStore$HMSHandler.createDefaultRoles_core(HiveMetaStore.java:656)
            at org.apache.hadoop.hive.metastore.HiveMetaStore$HMSHandler.createDefaultRoles(HiveMetaStore.java:648)
            at org.apache.hadoop.hive.metastore.HiveMetaStore$HMSHandler.init(HiveMetaStore.java:462)
            at org.apache.hadoop.hive.metastore.HiveMetaStore$HMSHandler.<init>(HiveMetaStore.java:419)
            at org.apache.hadoop.hive.metastore.HiveMetaStore$HMSHandler.<init>(HiveMetaStore.java:412)
```
And the stack trace matches hive 1.2.1:
https://github.com/apache/hive/blob/release-1.2.1/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java#L462

Reply:
If we are really using Hive 2.3.x and then Hive 1.2.1 randomly then this is a big issue. We should make this deterministic.

Reply:
Sorry I just meant test results are not deterministic. A test may fail if it's sharing JVM and runs after some other tests. Then in another run, if it doesn't share JVM, or runs with different other tests, it can pass.
I'm not saying hive version is randomly chosen.

Reply:
So we reverted this part of the change, for the "global" derby dir approach